gitclone/steperror.go

```diff
@@ -1,11 +1,178 @@
+			`fatal: repository '(.+)' not found`:                                                     fetchFailedCouldNotFindGitRepository,
+			`fatal: '(.+)' does not appear to be a git repository`:                                   fetchFailedCouldNotFindGitRepository,
+			`fatal: (.+)/info/refs not valid: is this a git repository?`:                             fetchFailedCouldNotFindGitRepository,
+			`remote: HTTP Basic: Access denied[\n]*fatal: Authentication failed for '(.+)'`:          fetchFailedHTTPAccessError,
+			`remote: Invalid username or password\(\.\)[\n]*fatal: Authentication failed for '(.+)'`: fetchFailedHTTPAccessError,
+			`Unauthorized`:                          fetchFailedHTTPAccessError,
+			`Forbidden`:                             fetchFailedHTTPAccessError,
+			`remote: Unauthorized LoginAndPassword`: fetchFailedHTTPAccessError,
+			// `fatal: unable to access '(.+)': Failed to connect to .+ port \d+: Connection timed out
+			// `fatal: unable to access '(.+)': The requested URL returned error: 400`
+			// `fatal: unable to access '(.+)': The requested URL returned error: 403`
+			`fatal: unable to access '(.+)': (Failed|The requested URL returned error: \d+)`: fetchFailedHTTPAccessError,
+			// `ssh: connect to host (.+) port \d+: Connection timed out`
+			// `ssh: connect to host (.+) port \d+: Connection refused`
+			// `ssh: connect to host (.+) port \d+: Network is unreachable`
+			`ssh: connect to host (.+) port \d+:`:                                fetchFailedCouldConnectError,
+			`ssh: Could not resolve hostname (.+): Name or service not known`:    fetchFailedCouldConnectError,
+			`fatal: unable to access '.+': Could not resolve host: (\S+)`:        fetchFailedCouldConnectError,
+			`ERROR: The \x60(.+)' organization has enabled or enforced SAML SSO`: fetchFailedSamlSSOError,
+		})
+}
+
+func mapRecommendation(tag string, err error) step.Recommendation {
+	msg := err.Error()
+	switch tag {
+	case checkoutFailedTag:
+		matcher := newCheckoutFailedPatternErrorMatcher()
+		return matcher.Run(msg)
+	case updateSubmodelFailedTag: // update_submodule_failed could have the same errors as fetch
+		fallthrough
+	case fetchFailedTag:
+		fetchFailedMatcher := newFetchFailedPatternErrorMatcher()
+		return fetchFailedMatcher.Run(msg)
+	}
+	return nil
+}
 
 func newStepError(tag string, err error, shortMsg string) *step.Error {
+	recommendation := mapRecommendation(tag, err)
+	if recommendation != nil {
+		return step.NewErrorWithRecommendations("git-clone", tag, err, shortMsg, recommendation)
+	}
+
 	return step.NewError("git-clone", tag, err, shortMsg)
 }
 
+func newDetailedErrorRecommendation(detail Detail) step.Recommendation {
+	return step.Recommendation{
+		"DetailedError": map[string]string{
+			"Title":       detail.Title,
+			"Description": detail.Description,
+		},
+	}
+}
+
+func checkoutFailedGenericDetail(params ...string) Detail {
```

Comment:
these are constructors of `Detail`, i suggest using `new` prefix and `Detail` suffix for these contractors.

Reply:
Also by convention, first please define the struct, then the constructors, then struct functions.